### ISSUE TITLE
Inference differences between this fork and the standard pjreddie darknet

### ISSUE THREAD

**Issue by shehrzad** [AlexeyAB/darknet#4283]
Hello, big fan of your work here. I trained a YOLOv3 Full model for 10,000 iterations that seems to be working well. If I run the darknet demo on my test MP4 using my .cfg and .weights file, I tabulate the detections to a CSV file. That is, for each frame I list out each detected object along with the probability and bounding box. I inserted some simple code in the detect_in_thread() function to save this data off to a CSV file.

I then take the same .cfg, .weights, and MP4 files and invoke the darknet demo on the standard pjreddie darknet build. I've made sure **thresh** and **hier_thresh** are the same (0.5 in both cases). I would expect quite similar inference results, but that is not what I observe. The pjreddie darknet outputs significantly _fewer_ detected objects. If I inspect the results coming out of this code base more closely, it seems like I get double detections. More specifically, I get one bounding box of say class k with probability 0.92, and then another nearly similar bounding box of the same class k but with probability 0.68. I would expect the latter bounding box to be non-maxima suppressed. I feel like if this delta were addressed, the two code bases would perform more similarly (but still not identically).

Any ideas?

**Comment by AlexeyAB** [comment]
Try to use flag `-letter_box` at the end of detection command in this repository

**Comment by shehrzad** [comment]
Thanks @AlexeyAB , adding the `letter_box` command-line option certainly made things better, in terms of a tighter match between the standard pjreddie darknet. However I still remain concerned about duplicate detections. Consider this frame where the standard darknet finds two objects, one of class 5 and the other of class 2 with the (x,y,w,h) bounding box coordinates to the right of the class 5 and class 2 detection:

```
5, 0.961610, 0.447716, 0.377555, 0.356563, 0.407533
2, 0.906066, 0.675455, 0.539794, 0.231823, 0.208153
```

When I run the inference on this code I get _four_ detections, two each for class 5 and class 2:

```
5, 0.962566, 0.447471, 0.378103, 0.356619, 0.407124
5, 0.956374, 0.470948, 0.373279, 0.428596, 0.379566
2, 0.903827, 0.675454, 0.540227, 0.232432, 0.208529
2, 0.872524, 0.688035, 0.532464, 0.228656, 0.212197
```

It looks like the 1st class 5 detection and 1st class 2 detection correspond to the standard pjreddie darknet detection. And it feels like the other two are potential duplicates that should have been non-maxima suppressed.

**Comment by AlexeyAB** [comment]
Can you show output prediction.jpg file with bboxes of detected objects?

**Comment by shehrzad** [comment]
I started to write the code to save out the prediction.jpg file including the bounding boxes when I figured out the problem. I was saving my CSV file in the incorrect spot in the code. In contrast to the pjreddie code base, you do the NMS stuff (`do_nms_sort`) inside of the `demo` function. Whereas in the pjreddie code base, the possibly slightly different NMS stuff is performed right within the thread itself -- that is, within the function `detect_in_thread`. Therefore in your code I was saving the detected bboxes _prior_ to the non-maxima suppression.
